traffic_ops/client/serviceCategory.go

```diff
@@ -0,0 +1,133 @@
+	reqBody, err := json.Marshal(serviceCategory)
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+	resp, remoteAddr, err := to.request(http.MethodPost, API_SERVICE_CATEGORIES, reqBody)
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+	defer resp.Body.Close()
+	var alerts tc.Alerts
+	err = json.NewDecoder(resp.Body).Decode(&alerts)
+	return alerts, reqInf, nil
+}
+
+// UpdateServiceCategoryByID updates a service category by ID
+func (to *Session) UpdateServiceCategoryByID(id int, serviceCategory tc.ServiceCategory) (tc.Alerts, ReqInf, error) {
+
+	var remoteAddr net.Addr
+	reqBody, err := json.Marshal(serviceCategory)
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+	route := fmt.Sprintf("%s/%d", API_SERVICE_CATEGORIES, id)
+	resp, remoteAddr, err := to.request(http.MethodPut, route, reqBody)
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+	defer resp.Body.Close()
+	var alerts tc.Alerts
+	err = json.NewDecoder(resp.Body).Decode(&alerts)
+	return alerts, reqInf, nil
+}
+
+// GetServiceCategories returns a list of service categories
+func (to *Session) GetServiceCategories() ([]tc.ServiceCategory, ReqInf, error) {
+	resp, remoteAddr, err := to.request(http.MethodGet, API_SERVICE_CATEGORIES, nil)
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+	if err != nil {
+		return nil, reqInf, err
+	}
+	defer resp.Body.Close()
+
+	var data tc.ServiceCategoriesResponse
+	err = json.NewDecoder(resp.Body).Decode(&data)
+	return data.Response, reqInf, nil
+}
+
+// GetServiceCategoryByID gets a service category by the service category id
+func (to *Session) GetServiceCategoryByID(id int) ([]tc.ServiceCategory, ReqInf, error) {
+	route := fmt.Sprintf("%s?id=%d", API_SERVICE_CATEGORIES, id)
+	resp, remoteAddr, err := to.request(http.MethodGet, route, nil)
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+	if err != nil {
+		return nil, reqInf, err
+	}
+	defer resp.Body.Close()
+
+	var data tc.ServiceCategoriesResponse
+	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return nil, reqInf, err
+	}
+
+	return data.Response, reqInf, nil
+}
+
+// GetServiceCategoryByName gets a service category by the service category name
+func (to *Session) GetServiceCategoryByName(name string) ([]tc.ServiceCategory, ReqInf, error) {
+	url := fmt.Sprintf("%s?name=%s", API_SERVICE_CATEGORIES, url.QueryEscape(name))
+	resp, remoteAddr, err := to.request(http.MethodGet, url, nil)
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+	if err != nil {
+		return nil, reqInf, err
+	}
+	defer resp.Body.Close()
+
+	var data tc.ServiceCategoriesResponse
+	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return nil, reqInf, err
+	}
+
+	return data.Response, reqInf, nil
+}
```

Comment:
Per a recent mailing list discussion, this should instead just be one function that takes in optional `url.Values` (i.e. a `*url.Values` so that it can be `nil`) so that callers can filter as they please without needing to maintain multiple functions.